File: Analizador Semantico/Analizador/src/AST.java
import java.util.List;

public class GeneradorAST {
    private List<Token> postfija;
    private List<Nodo> stackk;
    private Postfix poshelp;

    public GeneradorAST(List<Token> postfija) {
        this.postfija = postfija;
        this.stackk = new ArrayList<>();
        this.poshelp = new Postfix(null);
    }

    public Arbol generarAST() {
        List<Nodo> stackPadres = new ArrayList<>();
        Nodo raiz = new Nodo(new Token(TokenType.NULL, "", "", null));
        stackPadres.add(raiz);
        Nodo padre = raiz;
        for (int i = 0; i < postfija.size(); i++) {
            Token t = postfija.get(i);
            if (t.getType() == TokenType.EOF) {
                break;
            }
            if (poshelp.palabras_reservadas.containsValue(t.getType())) {
                Nodo n = new Nodo(t);
                padre = stackPadres.get(stackPadres.size() - 1);
                padre.insertar_sig_hijo(n);
                stackPadres.add(n);
                padre = n;
            } else if (poshelp.esOperando(t.getType())) {
                Nodo n = new Nodo(t);
                stackk.add(n);
            } else if (poshelp.esOperador(t.getType())) {
                int aridad = poshelp.obAridad(t.getType());
                Nodo n = new Nodo(t);
                for (int j = 0; j < aridad; j++) {
                    Nodo nodoAux = stackk.remove(stackk.size() - 1);
                    n.insertar_hijo(nodoAux);
                }
                stackk.add(n);
            } else if (t.getType() == TokenType.SEMICOLON) {
                if (stackk.size() == 0) {
                    stackPadres.remove(stackPadres.size() - 1);
                    padre = stackPadres.get(stackPadres.size() - 1);
                } else {
                    Nodo n = stackk.remove(stackk.size() - 1);
                    if (padre.getValue().getType() == TokenType.VAR) {
                        if (n.getValue().getType() == TokenType.ASSIGNATION) {
                            padre.insertar_hijos(n.getHijos());
                        } else {
                            padre.insertar_sig_hijo(n);
                        }
                        stackPadres.remove(stackPadres.size() - 1);
                        padre = stackPadres.get(stackPadres.size() - 1);
                    } else if (padre.getValue().getType() == TokenType.PRINT) {
                        padre.insertar_sig_hijo(n);
                        stackPadres.remove(stackPadres.size() - 1);
                        padre = stackPadres.get(stackPadres.size() - 1);
                    } else {
                        padre.insertar_sig_hijo(n);
                    }
                }
            }
        }
        Arbol programa = new Arbol(raiz);
        return programa;
    }

    public void printAST(Nodo nodo, int nivel) {
        if (nodo == null) {
            return;
        }
        System.out.println(" ".repeat(nivel) + nodo.getValue());
        if (nodo.getHijos() != null) {
            for (Nodo i : nodo.getHijos()) {
                printAST(i, nivel + 1);
            }
        }
    }
}

File: Analizador Semantico/AST.py
from Tokentype import TokenType
from Token import Token
from Nodo import Nodo
from Postfija import Postfix
from Arbol import Arbol

class GeneradorAST:
    def __init__(self, postfija:list[Token]) -> None:
        self.postfija = postfija
        self.stackk= []
        self.poshelp = Postfix(None)

    def generarAST(self):
        stackPadres = []
        raiz = Nodo(Token(TokenType.NULL,"","",None))
        stackPadres.append(raiz)

        padre:Nodo = raiz

        for i,t in enumerate(self.postfija):
            #print(f"{t} y padre {padre}")
            if t.type == TokenType.EOF:
                break
            
            if t.type in self.poshelp.palabras_reservadas.values():
                n = Nodo(t)
                padre = stackPadres[-1]
                padre.insertar_sig_hijo(n)
                stackPadres.append(n)
                padre = n
            elif self.poshelp.esOperando(t.type):
                #print(f"{t.type}")
                n = Nodo(t)
                self.stackk.append(n)
            elif self.poshelp.esOperador(t.type):
                aridad = self.poshelp.obAridad(t.type)
                n = Nodo(t)
                for i in range(aridad):
                    nodoAux = self.stackk.pop()
                    n.insertar_hijo(nodoAux)
                self.stackk.append(n)
            elif t.type == TokenType.SEMICOLON:
                
                if len(self.stackk) == 0:
                    stackPadres.pop()
                    padre = stackPadres[-1]
                else:
                    n:Nodo = self.stackk.pop()

                    #print(f"{padre.value}")
                    if padre.value.type == TokenType.VAR:
                        if n.value.type == TokenType.ASIGNATION:
                            padre.insertar_hijos(n.hijos)
                        else:
                            padre.insertar_sig_hijo(n)
                        stackPadres.pop()
                        padre = stackPadres[-1]
                    elif padre.value.type == TokenType.PRINT:
                        padre.insertar_sig_hijo(n)
                        stackPadres.pop()
                        padre = stackPadres[-1]
                    else:
                        padre.insertar_sig_hijo(n)

        #self.printAST(raiz)
        programa = Arbol(raiz)
        return programa
    
    def printAST(self, nodo:Nodo, nivel=0):
        if nodo is None:
            return
        
        print(f"{' '*nivel}{nodo.value}")
        
        if nodo.hijos:
            for i in nodo.hijos:
                self.printAST(i,nivel+1)
        

                        
            

            



    